src/sql/workbench/parts/objectExplorer/browser/iconRenderer.ts

```diff
@@ -0,0 +1,371 @@
+/*---------------------------------------------------------------------------------------------
+ *  Copyright (c) Microsoft Corporation. All rights reserved.
+ *  Licensed under the Source EULA. See License.txt in the project root for license information.
+ *--------------------------------------------------------------------------------------------*/
+
+import { generateUuid } from 'vs/base/common/uuid';
+
+export function renderServerIcon(element: HTMLElement, serverInfo: ServerInfo, isConnected: boolean,
+	iconSizePx?: number, topMarginPx?: number, leftMarginPx?: number
+): void {
+	let serverIcon: IconInfo = ServerIcon.getServerIcon(serverInfo);
+	renderServerIconByPath(element, serverIcon, isConnected, iconSizePx, topMarginPx, leftMarginPx);
+}
+
+export function renderServerIconByPath(element: HTMLElement, iconData: IconPath | IconInfo, isConnected: boolean,
+	iconSizePx?: number, topMarginPx?: number, leftMarginPx?: number
+): void {
+	if (!element) { return; }
+
+	let iconPath: IconPath = undefined;
+	if (iconData['path']) {
+		let iconInfo = iconData as IconInfo;
+		iconPath = iconInfo.path;
+		iconSizePx = iconSizePx || Math.min(iconInfo.widthPx, iconInfo.heightPx);
+	} else {
+		iconPath = iconData as IconPath;
+	}
+
+	renderIcon(element, iconPath);
+	let badge: ServerStatusBadge = new ServerStatusBadge(iconSizePx, topMarginPx, leftMarginPx);
+	badge.renderBadge(element, isConnected);
+}
+
+export function renderIcon(element: HTMLElement, iconPath: string | IconPath): void {
+	if (!element || !iconPath) { return; }
+
+	let lightPath: string = undefined;
+	let darkPath: string = undefined;
+	if (typeof iconPath === 'string') {
+		lightPath = iconPath;
+		darkPath = iconPath;
+	} else {
+		lightPath = iconPath.light;
+		darkPath = iconPath.dark;
+	}
+
+	if (element.id === undefined || element.id === '') {
+		element.id = `id_${generateUuid()}`;
+	}
+	let elementId: string = element.id;
+	let styleId: string = `icon_style_${elementId}`;
+
+	let current = element.previousElementSibling;
+	while (current) {
+		let next = current.previousElementSibling;
+		if (current.id === styleId) {
+			current.remove();
+			break;
+		}
+		current = next;
+	}
+
+	element.insertAdjacentHTML('beforebegin',
```

Comment:
I think Alan mentioned this offline, but if we do this like here it'll create new CSS elements for every single time we show a server icon. The ideal is (I think) to define once & reuse. E.g. define the CSS for it globally and cache it. The modelview icon logic does this.

Reply:
fixed.